train/train.py

```diff
@@ -0,0 +1,196 @@
+    '''
+    # Input: [b, 2, N_FRAME_TOKS]
+    # Flattened: [b, 256]
+
+    # Bottleneck: [b, N_FRAME_TOKS + s] : s < 128 n Bottleneck[:128] = Input[:128]
+    # I.e. Transformation_Code = Encoder(Input): [b, s]
+    #      Output = Decoder(Input[:N_FRAME_TOKS] + Transformation_Code): [b, 128]
+
+    # Output: [b, 2, N_FRAME_TOKS]
+
+    '''
+    For now try to get results with constant bottleneck
+    i.e. try to get some encoder model that can encode the diffs in 50% less tokens
+
+    Next try variable length codings based on difference between tokens of frames
+    Need to add delimiter tokens <|X1|> abcd <|F|> 12 <|X2>| xzyw <|EOT|> pad pad pad 
+
+
+    Current Objective:
+        Get the model to learn the spatial embedding table by passing it through the dynamics
+        bottleneck
+
+        N_DYN_TOKS == N_SPATIAL_TOKS
+
+    '''
+
+    # Logging
+    enable_wandb = False
+
+    if enable_wandb:
+        wandb.init(
+            project="diff-encoding",
+        )
+
+    # Data Prep
+    device = "cuda" if torch.cuda.is_available() else "cpu"
+    spatial_embeddings = torch.load("embedding.pt").to(device)
+    spatial_embeddings.requires_grad = False
+
+    batch_size = 64
+    n_frames = 2
+    dataloader = TokenLoader('commavq-mini.npy', batch_size, n_frames=n_frames)
+
+    # Model Prep
+    N_DYNAMICS_TOKS = 64 # s = N_FRAME_TOKS - 64 = 128 / 2
+
+    enc = Encoder(
+        width=256,
+        layers=8,
+        heads=8,
+        n_tokens=N_DYNAMICS_TOKS,
+        n_input_tokens=n_frames*128 + n_frames,
+        spatial_embeddings=spatial_embeddings,
+    ).to(device)
+    dec = Decoder(
+        width=256,
+        layers=8,
+        heads=8,
+        n_tokens=N_DYNAMICS_TOKS,
+        n_input_tokens=N_DYNAMICS_TOKS + N_FRAME_TOKS + 2,
+        spatial_embeddings=spatial_embeddings,
+    ).to(device)
+    '''
+    q = Quantizer(
+        n_embeddings=128,
+        embedding_dim=256,
+        commitment_cost=0.25,
+    ).to(device)
+    '''
+
+    # Opt Prep
+    iters = 10000000
+
+    # opt = optim.AdamW(list(enc.parameters()) + list(dec.parameters()) + list(q.parameters()))
+    opt = optim.AdamW(list(enc.parameters()) + list(dec.parameters()))
+
+    i = 0
+    t0 = time.time()
+    for X in dataloader:
+        if i >= iters:
+            break
+        X = X.long().to(device)
+
+        data_time = time.time() - t0
+
+        embs = spatial_embeddings[X].reshape(X.shape[0], X.shape[1], -1, spatial_embeddings.shape[-1])
+        e0 = embs[:, 0]
+        X0 = X[:, 0].reshape(X.shape[0], -1).long()
+        labels = X[:, 1:].reshape(X.shape[0], -1)
+
+        # Forward pass
+        opt.zero_grad()
+
+        f_emb = enc(X)
+
+        # f, ppl, encodings = q(f_emb)
+        f = f_emb
+
+        logits = dec(e0, f)
+
+        # TODO: why does this matter???
+        true_logits = logits[:, :N_FRAME_TOKS]
+        # true_logits = logits[:, -N_FRAME_TOKS:]
```

Comment:
this makes training either completely not work or work and I currently have no clue why